### PR TITLE
feat(traces): query spans into dataframes

### LINKED ISSUE #1548: [traces][rag] add helper functions to make trace dataframes a more usable format
Trace data. as dataframes can be difficult to use for data science as certain fields can be nested and the columns can be noisy. We have three primary types of records we need to morph traces into

- [ ] Trace or Generation Records with context (captures the input, output, context, ...)
- [ ] Document Records (captures input, document, offset)
- [ ] Trace or Generation Records (simple IO)

### PR DESCRIPTION
**Pull request by RogerHYang**
resolves #1548

### CODE REVIEW

File: src/phoenix/session/session.py [hunk 1]
```diff
@@ -167,6 +168,27 @@ def url(self) -> str:
         """Returns the url for the phoenix app"""
         return _get_url(self.host, self.port, self.notebook_env)
 
+    def query_spans(
+        self,
+        *queries: SpanQuery,
+        start_time: Optional[datetime] = None,
+        stop_time: Optional[datetime] = None,
+        root_spans_only: Optional[bool] = None,
```
Comment (mikeldking) [inline src/phoenix/session/session.py:176]: Seems slightly inconsistent that this is outside?

File: src/phoenix/trace/dsl/helpers.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        session.query_spans(
+            SpanQuery()
+            .select(**INPUT)
+            .explode(RETRIEVAL_DOCUMENTS, reference=DOCUMENT_CONTENT, score=DOCUMENT_SCORE)
+            .where(IS_RETRIEVER)
+        ),
+    )
+
+
+def get_qa_with_reference(session: Session) -> pd.DataFrame:
```
Comment (mikeldking) [inline src/phoenix/trace/dsl/helpers.py:40]: the word reference always trips me up - so used to it being context.

File: src/phoenix/trace/dsl/query.py [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+    def _from_context(span: Span, key: str) -> Any:
+        return getattr(span.context, key, None)
+
+    @staticmethod
+    def _from_span(span: Span, key: str) -> Any:
+        return getattr(span, key, None)
+
+
+@dataclass(frozen=True)
+class Explosion(Projection):
```
Comment (mikeldking) [inline src/phoenix/trace/dsl/query.py:69]: Don't forget to write down you ideas in docstrings.

File: src/phoenix/trace/dsl/query.py [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+            )
+            if not record:
+                continue
+            record[self.primary_index.key] = self.primary_index(span)
+            record[f"{self.position_prefix}position"] = i
+            yield record
+
+
+@dataclass(frozen=True)
+class Concatenation(Projection):
```
Comment (mikeldking) [inline src/phoenix/trace/dsl/query.py:125]: Cam as above, don't forget to define your DSL

File: src/phoenix/trace/dsl/helpers.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+IS_LLM = "span_kind == 'LLM'"
+IS_RETRIEVER = "span_kind == 'RETRIEVER'"
+
+
+class Session(Protocol):
+    def query_spans(self, *query: SpanQuery) -> Optional[Union[pd.DataFrame, List[pd.DataFrame]]]:
+        ...
+
+
+def get_retrieved_documents(session: Session) -> pd.DataFrame:
```
Comment (mikeldking) [inline src/phoenix/trace/dsl/helpers.py:28]: need to be exported at a lower level in the module tree `trace.dsl.helpers` is too deep. Maybe session evals like you have

**Review by mikeldking — APPROVED** [review]

**Review by axiomofjoy — APPROVED** [review]

File: src/phoenix/session/session.py (hunk 1 above)
Comment (RogerHYang) [reply src/phoenix/session/session.py:176]: yea, let me remove it

File: src/phoenix/trace/dsl/helpers.py (hunk 2 above)
Comment (RogerHYang) [reply src/phoenix/trace/dsl/helpers.py:40]: in this case it's referring to the template variable `reference`

File: src/phoenix/trace/dsl/query.py (hunk 3 above)
Comment (RogerHYang) [reply src/phoenix/trace/dsl/query.py:69]: yup, I will do a fast-follow.

File: src/phoenix/trace/dsl/helpers.py (hunk 5 above)
Comment (RogerHYang) [reply src/phoenix/trace/dsl/helpers.py:28]: ok, i'll just re-export it over there for now before we find a final home

**Pull request merged**